### DISCUSSION TITLE
Installation Issue

### DISCUSSION THREAD

**Discussion by roeberp** [alexjustesen/speedtest-tracker#651] [Q&A]
Hey everyone,

I'm trying to install this on a Proxmox server running a Docker container. My run command is the following:

docker run -d --name speedtest-tracker --restart unless-stopped \
    -p 8080:80 \
    -p 8443:443 \
    -e PUID=1000 \
    -e PGID=1000 \
    -v /docker/speedtest-tracker/data:/config \
    -v /docker/speedtest-tracker/web:/etc/ssl/web \
    ghcr.io/alexjustesen/speedtest-tracker:latest

When I run this, I get an error and have had no lucking finding this exact error on my many searches.

docker: failed to register layer: unlinkat /usr/share/doc/adduser/examples/adduser.local.conf.examples: invalid argument.

If anyone has any suggestions to resolve this, I would be very grateful! I'm fairly new to docker and am still learning. This one has me stumped.

Thanks!

**Comment by Ramalama2** [comment]
```
latest: Pulling from alexjustesen/speedtest-tracker
2ab09b027e7f: Pull complete
9a37c607e3b8: Pull complete
e516956e04f5: Pull complete
015f5e2d3346: Pull complete
affb68e809e1: Pull complete
4d0fd514210a: Extracting [==================================================>]   37.4MB/37.4MB
2896d6351935: Download complete
944b1a76708e: Download complete
cc762f8fceb1: Download complete
7c87d6420934: Download complete
4b292aa4d561: Download complete
9b3cbd24ae8d: Download complete
1e116dfd9a0c: Download complete
923769cc472c: Download complete
ff2ddad42278: Download complete
fddf580d9660: Download complete
369965f31669: Download complete
92b2acc1b9e2: Download complete
4f4fb700ef54: Download complete
dc34912bf347: Download complete
1c6082b02786: Download complete
failed to register layer: unlinkat /usr/share/doc/adduser/examples/adduser.local.conf.examples: invalid argument
Unable to find image 'ghcr.io/alexjustesen/speedtest-tracker:latest' locally
latest: Pulling from alexjustesen/speedtest-tracker
```

Same here, seems like the issue is inside the 4d0fd514210a image.

**Comment by Ramalama2** [comment]
> Hey everyone,
> 
> I'm trying to install this on a Proxmox server running a Docker container. My run command is the following:
> 
> docker run -d --name speedtest-tracker --restart unless-stopped -p 8080:80 -p 8443:443 -e PUID=1000 -e PGID=1000 -v /docker/speedtest-tracker/data:/config -v /docker/speedtest-tracker/web:/etc/ssl/web ghcr.io/alexjustesen/speedtest-tracker:latest
> 
> When I run this, I get an error and have had no lucking finding this exact error on my many searches.
> 
> docker: failed to register layer: unlinkat /usr/share/doc/adduser/examples/adduser.local.conf.examples: invalid argument.
> 
> If anyone has any suggestions to resolve this, I would be very grateful! I'm fairly new to docker and am still learning. This one has me stumped.
> 
> Thanks!

I think youre using docker inside an unpriviliged container on proxmox.
At least that would explain our issue and what i can find on google.

EDIT:
Good News: Once OpenZFS 2.2 arrives to Proxmox, we won't have such issues anymore, but that will take long, i think 3-4 Months still.
Otherwise im trying to find a solution on google xD

**Comment by Ramalama2** [comment]
Okay, in my digging research:

1. i've downloaded the image files with skopeo:
`skopeo copy docker://ajustesen/speedtest-tracker:latest dir:asd`

2. I've made a small script, to check the uid:gid owners of the files inside the tar archives. Because i wanted to check if this issue is the same as ours: https://github.com/moby/moby/issues/43576
```
#!/bin/bash

for file in *; do
    # Ensure we're working with a file and it's not a .txt file
    if [[ -f "$file" && "${file##*.}" != "txt" && "${file##*.}" != "sh" ]]; then
            # Cut the filename to the first 12 characters
        output="${file:0:12}.txt"

        # Execute the command and save output
        tar tvf "$file" > "$output"
    fi
done
```

3. the files are attached:
[1e116dfd9a0c.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963291/1e116dfd9a0c.txt)
[2ab09b027e7f.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963292/2ab09b027e7f.txt)
[4b292aa4d561.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963293/4b292aa4d561.txt)
[4d0fd514210a.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963294/4d0fd514210a.txt)
[4f4fb700ef54.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963295/4f4fb700ef54.txt)
[7c87d6420934.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963296/7c87d6420934.txt)
[9a37c607e3b8.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963297/9a37c607e3b8.txt)
[9b3cbd24ae8d.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963298/9b3cbd24ae8d.txt)
[015f5e2d3346.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963299/015f5e2d3346.txt)
[91b8f98bf53c.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963300/91b8f98bf53c.txt)
[157ef81a5fdd.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963301/157ef81a5fdd.txt)
[330e84f00c77.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963302/330e84f00c77.txt)
[920bad91bc9c.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963303/920bad91bc9c.txt)
[944b1a76708e.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963304/944b1a76708e.txt)
[2896d6351935.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963305/2896d6351935.txt)
[923769cc472c.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963306/923769cc472c.txt)
[094417573c89.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963307/094417573c89.txt)
[affb68e809e1.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963308/affb68e809e1.txt)
[cc762f8fceb1.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963309/cc762f8fceb1.txt)
[e516956e04f5.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963310/e516956e04f5.txt)
[fddf580d9660.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963311/fddf580d9660.txt)
[ff2ddad42278.txt](https://github.com/alexjustesen/speedtest-tracker/files/11963312/ff2ddad42278.txt)

It's very late now, i have to sleep, maybe someone else can look through for unusual gid/uid or whatever else.
Maybe deleting the folder "/usr/share/doc/adduser/examples/adduser.local.conf.examples" from the image would fix it anyway.
Cheers

**Comment by Ramalama2** [comment]
okay further digging, the text files and the gid/uid's seems okay.
There is nothing unusual, the only thing comes into my mind is, that maybe the "path" is simply too long or sth like that.

tbh, it makes no sense to me why speedtest-traker doesn't work, while every other docker image im using works flawless.
```
fde8b6826ca1   mariadb:latest                         "docker-entrypoint.s…"   13 hours ago   Up 8 minutes             3306/tcp                                                                   mariadb
f4841d97da24   portainer/portainer-ce                 "/portainer"             14 hours ago   Up 8 minutes             8000/tcp, 9000/tcp, 9443/tcp                                               portainer
8ad839861887   onlyoffice/documentserver:latest       "/app/ds/run-documen…"   14 hours ago   Up 8 minutes             80/tcp, 443/tcp                                                            onlyoffice
f474275243fa   vaultwarden/server:latest              "/start.sh"              14 hours ago   Up 8 minutes (healthy)   80/tcp, 3012/tcp                                                           vaultwarden
307e6db87586   nextcloud:latest                       "/entrypoint.sh apac…"   14 hours ago   Up 8 minutes             80/tcp                                                                     nextcloud
05228a48ecd7   nginx:latest                           "/docker-entrypoint.…"   14 hours ago   Up 8 minutes             80/tcp                                                                     nginx
dea4a5d0929a   redis:latest                           "docker-entrypoint.s…"   14 hours ago   Up 8 minutes             6379/tcp                                                                   redis
a1de2bcd42f1   linuxserver/heimdall:latest            "/init"                  14 hours ago   Up 8 minutes             80/tcp, 443/tcp                                                            heimdall
dd5e96cdaa8a   humenius/traefik-certs-dumper:latest   "/usr/bin/dump"          14 hours ago   Up 8 minutes (healthy)                                                                              cert-dumper
1120163b5f2a   authelia/authelia                      "/app/entrypoint.sh …"   14 hours ago   Up 8 minutes (healthy)   9091/tcp                                                                   authelia
9d2f046fa14c   traefik:latest                         "/entrypoint.sh trae…"   14 hours ago   Up 8 minutes             0.0.0.0:80->80/tcp, :::80->80/tcp, 0.0.0.0:443->443/tcp, :::443->443/tcp   proxy
```

That are the running containers, that have no issues whatsoever.

im gonna check as next if there is an adduser example folder either.
If they don't contain it, then im suspecting that the path is simply too long, because overlay2 and zfs aren't 100% compatible until openzfs 2.2.

However, the workaround for this situation would be atm, to make either an zfs dataset and format it as ext4, then run the LXC Container with docker on that storage.
Or making an VM for now.
The only sad thing about this, this workaround is only needed for speedtest-tracker, at least here.

However, if i know more, ill post.
Cheers

**Comment by Ramalama2** [comment]
Okay, after further digging:
traefik: doesn't have an /usr/share/doc/ folder at all.
authelia: doesn't have an /usr/share/doc/ folder at all.
cert-dumper: doesn't have an /usr/share/doc/ folder at all.
heimdall: doesn't have an /usr/share/doc/ folder at all.
redis: has only one file: /usr/share/doc/adduser/copyright
nginx: has only one file: /usr/share/doc/adduser/copyright
nextcloud: has only one file: /usr/share/doc/adduser/copyright
vaultwarden: has only one file: /usr/share/doc/adduser/copyright

onlyoffice: has all the files??:
```
root@8ad839861887:/# tree /usr/share/doc/adduser/
/usr/share/doc/adduser/
├── changelog.gz
├── copyright
└── examples
    └── adduser.local.conf.examples
        ├── skel
        └── skel.other

4 directories, 2 files
```

inspecting onlyoffice image with tar:
```
drwxr-xr-x 0/0               0 2023-05-22 14:04 usr/share/doc/
drwxr-xr-x 0/0               0 2023-05-22 14:04 usr/share/doc/adduser/
-rw-r--r-- 0/0            1864 2021-01-06 06:16 usr/share/doc/adduser/changelog.gz
-rw-r--r-- 0/0            1995 2018-09-15 22:14 usr/share/doc/adduser/copyright
drwxr-xr-x 0/0               0 2023-05-22 14:04 usr/share/doc/adduser/examples/
drwxr-xr-x 0/0               0 2023-05-22 14:04 usr/share/doc/adduser/examples/adduser.local.conf.examples/
drwxr-xr-x 0/0               0 2023-05-22 14:04 usr/share/doc/adduser/examples/adduser.local.conf.examples/skel/
drwxr-xr-x 0/0               0 2023-05-22 14:04 usr/share/doc/adduser/examples/adduser.local.conf.examples/skel.other/
```

comparing with speedtest-tracker:
```
drwxr-xr-x 0/0               0 2023-03-08 02:05 usr/share/doc/
drwxr-xr-x 0/0               0 2023-03-08 02:05 usr/share/doc/adduser/
-rw-r--r-- 0/0            1864 2021-01-06 06:16 usr/share/doc/adduser/changelog.gz
-rw-r--r-- 0/0            1995 2018-09-15 22:14 usr/share/doc/adduser/copyright
drwxr-xr-x 0/0               0 2023-03-08 02:05 usr/share/doc/adduser/examples/
drwxr-xr-x 0/0               0 2023-03-08 02:05 usr/share/doc/adduser/examples/adduser.local.conf.examples/
drwxr-xr-x 0/0               0 2023-03-08 02:05 usr/share/doc/adduser/examples/adduser.local.conf.examples/skel/
drwxr-xr-x 0/0               0 2023-03-08 02:05 usr/share/doc/adduser/examples/adduser.local.conf.examples/skel.other/
```

I don't see a single difference....
Im out of ideas, no clue why onlyoffice works and speedtest-tracker don't....
I can delete/pull everything with onlyoffice without issues, but getting always the error with speedtest-tracker.

the only other possibility here is, that simply the error from docker is maybe completely wrong, maybe it fails before or after and just outputs the last file it was doing, or some before, who knows.
At least im out.
Have no ideas and nothing i can try out.

Maybe someone else has more Luck in finding the issue/solution to it.

Cheers

**Comment by alexdelprete** [comment]
You are spending all these efforts but both are failing to provide BASIC INFO: are you using docker in a LXC container or VM? Are you using ZFS?

Docker is not officially supported in Proxmox LXC containers with ZFS, but they work in latest versions only, with latest kernels and only if privileged LXCs.

So don't waste time checking docker images, they are perfectly fine, it's your environment.

**Comment by alexdelprete** [comment]
This is my Debian 12 LXC container for Docker on Proxmox 8.x and ZFS, you need to properly configure it (check red rectangle):

![image](https://github.com/alexjustesen/speedtest-tracker/assets/7027842/57bd3972-23e8-4e50-8ae8-1e6d066b4ed5)

**Comment by Ramalama2** [comment]
Bro, i have multiple Proxmox Servers, almost all with ZFS Exclusive Storage, one of them has an EXT4 Storage tho.
On everyone is an Unpriviliged LXC Container with Docker inside.

- This Server has an EXT4 Storage, Unpriviliged with Nesting, Speedtest-tracker works perfectly fine:
<img width="733" alt="Bildschirmfoto 2023-07-06 um 18 29 23" src="https://github.com/alexjustesen/speedtest-tracker/assets/6314556/b949732f-ec45-445c-9dff-30e1d7f192ea">

![EXT4-Storage](https://github.com/alexjustesen/speedtest-tracker/assets/6314556/9f1ef18c-cf59-413a-b499-394bd1394529)

- This Server has an ZFS Storage, Unpriviliged with Nesting, every Docker Image except Speedtest-tracker works perfectly fine:
<img width="742" alt="Bildschirmfoto 2023-07-06 um 18 29 58" src="https://github.com/alexjustesen/speedtest-tracker/assets/6314556/e157f45d-26b8-424c-8d88-76215affc42a">

![ZFS-Storage](https://github.com/alexjustesen/speedtest-tracker/assets/6314556/a2850cbb-9eee-4c59-bd99-b2c3e2bf1f40)

My second post in this "Discussion" mentions already Proxmox in an Unpriviliged Container.
Running Docker in a Priviliged Container, is by no means the proper way, since it means basically the same as running Docker on the proxmox host itself and opening the ports on the firewall to the public.
Tho an Unpriviliged container is not very secure either, but however, its still a lot more secure.
A proper way would be running Docker inside an VM, but thats something that no one of us wants, due to resource/performance penalties xD

However, i was trying to find out why every other Docker image works with an unpriviliged container on zfs, while speedtest-tracker don't.
Even more mysterious is why Onlyoffice works, since its basically identical to speedtest-tracker.

However, it's obviously not only a fault of the Unpriviliged Container or "My Environment", since that would mean, that no other Docker image would work either.

**Comment by alexdelprete** [comment]
> Running Docker in a Priviliged Container, is by no means the proper way, since it means basically the same as running Docker on the proxmox host itself and opening the ports on the firewall to the public.

Yo, Bro...wazzuuup...:)

Jokes apart: I have many dockers on Proxmox, the critical ones are in VMs, because Proxmox DOES NOT SUPPORT Docker in LXC containers. Just read Proxmox official documentation. If you have an issue with Docker in LXC, there's no way to have OFFICIAL SUPPORT. Read their forums, there are a lot of issues open with docker in LXC, and the answer is always the same.

The only way I've managed to make docker work reliably in LXC is in privileged containers. But that is not supported either, so I'm on my own in case of problems. And no, it is not like running docker on Proxmox itself, it's a separate environment luckily. It's true that from a security standpoint it's not recommended, but it was the only way to have separation and make all the apps I needed working fine. Acceptable compromise for me.

> Even more mysterious is why Onlyoffice works, since its basically identical to speedtest-tracker.

No they're not identical. ST uses other images, it's based on s6, it's a laravel framework app (PHP). OnlyOffice is a .NET app IIRC, but you can't say two apps are IDENTICAL. Analyze them first.

> However, it's obviously not only a fault of the Unpriviliged Container or "My Environment", since that would mean, that no other Docker image would work either.

No, that is a totally wrong assumption. You don't use ALL apps/services that are out there. For example, I was having issues with Piwigo...all my apps were working, but Piwigo wasn't working, it threw weird errors when starting it up with docker compose. After a lot of tries, I set it up in my main docker that is in a VM, and had ZERO issues.

So, bottom line is: you are trying to make something work that is NOT SUPPORTED in the Proxmox environment, if it works fine, but if it doesn't, don't try to blame Proxmox, the app, the image or whatever, ok? That is why it's not supported, because it's not 100% compatible. I'm pretty sure in the future they'll make it work fine so I can finally use an unprivileged LXC, but I'll have to wait...

**Comment by Ramalama2** [comment]
Okay bro 😂

I never blamed anyone, either Proxmox or the Dev of Speedtest-tracker in any of the posts.

As you seen, im not even asked anybody to help and tryed to figure out myself, what the issue could be.

The whole discussion long, i tryed to figure out what's different, by myself, until i wasn't successful, since onlyoffice throw all my assumptions away 😂

So it's actually not clear to me, what you're trying to say other as it doesn't work on Unprivileged container and we don't have any support.

To be clear, if i wanted any support, or would assume that this is a bug of Speedtest-tracker, i would have opened a GitHub issue...
Instead of talking with myself in a discussion 😂

The Proxmox documentation is a bit slow, since they still recommend vfio modules that doesn't exist anymore on kernel 6.2, just to mention one if the documentation issues.
Overlay2 is already well supported on zfs 2.1, just not with all features, like renameat2.

That feature completeness will happen with zfs 2.2, probably in 1-3 Months.
Till zfs 2.2 will get out of RC.
The Proxmox team on the other hand is extremely fast with adopting updates.
With the exceptions of kernel opt-ins.
Incremental kernel updates and openzfs updates arrives to us almost instantly, maybe with some days delay only.

So as you see, im not blaming anyone 😂

Cheers

**Comment by alexdelprete** [comment]
No problem bro, just wanted to clear some things out. Love your enthusiasm, but what I'm telling you is that until Docker in LXC + ZFS will be officially supported by Proxmox, don't waste too much time and energy trying to make it work. Trust me, I've been there, done that. :)

I remember when ZFS could be used in LXC with Docker only through FUSE...then all of the sudden, with recent versions of edge kernels 6.x (before 8.x was out), overlay2 with ZFS started working fine also without FUSE (only in privileged LXCs).

There's my post on their forum regarding LXC+ZFS+Docker overlay2, and devs couldn't tell me HOW it was working. I thought it had to be the new kernel versions or something, but they said they didn't know exactly what made it work. Obviously, privileged LXC.

There are many things going on with all these integrations man...we can't control everything and try to understand why something doesn't work and others do...sometimes even devs don't have a clue. :D

But that's the fun part of having a homelab...tinkering...:)

Cheers mate.

**Comment by Ramalama2** [comment]
Yeah, I've been there either everywhere 😂
Docker inside lxc containers was always a fun part through the Proxmox history 😂

But somehow i got it always working on my part 😂
First with fuse a long time ago, then i made ext4 partitions always for docker only, then privileged containers with keyctl and nesting, and now Unprivileged with nesting only and without apparmor 😂

However, hopefully that pain will be all and for once over with openzfs 2.2 😂

Cheers

**Comment by roeberp** [comment]
Thanks for the help guys! It's really appreciated. I wish there was a better answer, but I guess at this point, I just need to wait or rebuild my Docker instance. Thanks again!

**Reply by Ramalama2** [reply]
There is a workaround to get it working, but it's määh...
And we all knew it probably since beginning...

Simply create a temporary dataset,
format it as ext4,
create a temporary folder on that dataset,
give the folder all rights, like "chown 100000:100000 folder" & "chmod 777 folder"
mount "mp0" that folder to your LXC Container, to any directory inside the lxc container, like /myfolder
stop & disable docker, (docker.service and docker.target)
rsync -av /var/lib/docker /myfolder
delete /var/lib/docker entirely & shutdown
change the mountpath mp0 to /var/lib/docker in your lxc config...
boot, start docker, pull speedtest-tracker etc...
stop docker again, make a temp folder like /mynewdocker
rsync -av /var/lib/docker /mynewdocker
shutdown the vm, remove the mountpath from the lxc config, destroy your ext4 temporary dataset...
start the vm, mv /mynewdocker to /var/lib/docker
start & enable all the services again (docker.service & docker.target)

it could be that you need to uninstall apparmor in your lxc container, im doing it anyway always...

I think that after that you can even pull/update newer speedtest-tracker images, because the directorys somehow already exist.

however, that solution is just shit, thats why i tryed to find a more "proper" way xD

Cheers

**Comment by Glitch3dPenguin** [comment]
It's a few months later and this is still an issue. I too normally use LXC containers and try my best to avoid using docker at all if possible. I figured I'd give docker a shot on this go around and I got the same issue as OP. I have plenty of other unprivileged containers using docker but as @alexdelprete had mentioned there is no direct support. It looks like you are either running this in your Docker VM if you are using Proxmox or doing the manual install.
